src/gui/qgsdatadefinedbutton.h

```diff
@@ -167,6 +181,12 @@ class GUI_EXPORT QgsDataDefinedButton: public QToolButton
     void clearCheckedWidgets() { mCheckedWidgets.clear(); }
 
     /**
+     * Set an assistant to define expression
+     * takes ownership
+     */
+    void setAssistant( QgsDataDefinedAssistant * assistant ) { mAssistant.reset( assistant ); }
```

Comment:
probably the parent of mAssistant should be set to "this" here too. That will help ensure that the dialog gets displayed in the right position (eg for multi monitor setups)